DemoAppPush/NotificationService.swift

```diff
@@ -0,0 +1,115 @@
+import StreamChat
+import UserNotifications
+
+class NotificationService: UNNotificationServiceExtension {
+    var contentHandler: ((UNNotificationContent) -> Void)?
+    var request: UNNotificationRequest?
+
+    func addAttachments(
+        url: URL,
+        content: UNMutableNotificationContent,
+        identifier: String = "image",
+        completion: @escaping (UNMutableNotificationContent) -> Void
+    ) {
+        let task = URLSession.shared.downloadTask(with: url) { (downloadedUrl, _, _) in
+            defer {
+                completion(content)
+            }
+
+            guard let downloadedUrl = downloadedUrl else {
+                return
+            }
+          
+            guard let path = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true).first else {
+                return
+            }
+          
+            let localURL = URL(fileURLWithPath: path).appendingPathComponent(url.lastPathComponent)
+          
+            do {
+                try FileManager.default.moveItem(at: downloadedUrl, to: localURL)
+            } catch {
+                return
+            }
+
+            guard let attachment = try? UNNotificationAttachment(identifier: identifier, url: localURL, options: nil) else {
+                return
+            }
+
+            content.attachments = [attachment]
+        }
+        task.resume()
+    }
+
+    func addMessageAttachments(
+        message: ChatMessage,
+        content: UNMutableNotificationContent,
+        completion: @escaping (UNMutableNotificationContent) -> Void
+    ) {
+        if let imageURL = message.author.imageURL {
+            addAttachments(url: imageURL, content: content) {
+                completion($0)
+            }
+            return
+        }
+        if let attachment = message.imageAttachments.first {
+            addAttachments(url: attachment.imageURL, content: content) {
+                completion($0)
+            }
+            return
+        }
+    }
```

Comment:
This also seems like helpers functions for the `UNNotificationServiceExtension` right? Can we move them to the bottom then? And I guess make them private? Are they used outside of the NotificationService? 

Another option is to extract these functions into a different component, it would allow testing `NotificationService` easier since these helper functions would be mockable. Maybe introducing a `NotificationContentService`.

Reply:
this is part of the demo, I did not want to expand the scope of the SDK and add more helpers. This serves as a decent example / starting point